src/timepicker/TimePicker.ts

```diff
@@ -0,0 +1,481 @@
+	step?: number; // defaults to 60
+	value?: string;
+}
+
+// Enum used when traversing items using arrow keys
+export const enum Operation {
+	increase = 1,
+	decrease = -1
+};
+
+// This can go away when Safari decides to support KeyboardEvent.key
+const keys: {[key: string]: number} = {
+	escape: 27,
+	enter: 13,
+	up: 38,
+	down: 40
+};
+
+function getTime(units: TimeUnits, date = new Date()) {
+	const { hour, minute, second } = units;
+	date.setHours(hour);
+	date.setMinutes(minute as number);
+	date.setSeconds(second as number);
+	date.setMilliseconds(0);
+	return date.getTime();
+}
+
+/**
+ * Convert a standard time string into an object with `hour`, `minute`, and `second` number properties.
+ *
+ * For example, '12:30' is converted to `{ hour: 12, minute: 30, second: 0 }`, and '19:03:27' is converted
+ * to `{ hour: 19, minute: 3, second: 27 }`.
+ *
+ * @param value   A standard time string or an object with `hour`, `minute`, and `second` properties.
+ * @throws Error  For any string not in the format 'HH:mm' or 'HH:mm:ss'.
+ * @return        An object containing `hour`, `second`, and `number` properties.
+ */
+export const parseUnits = (function () {
+	const TIME_PATTERN = /^\d{2}:\d{2}(:\d{2})?$/;
+	return function (value: string | TimeUnits): TimeUnits {
+		if (typeof value === 'string') {
+			if (!TIME_PATTERN.test(value)) {
+				throw new Error('Time strings must be in the format HH:mm or HH:mm:ss');
+			}
+
+			const [ hour, minute, second = 0 ] = value.split(':').map(unit => parseInt(unit, 10));
+			return { hour, minute, second } as TimeUnits;
+		}
+
+		return value;
+	};
+})();
+
+const TimePickerBase = ThemeableMixin(WidgetBase);
+
+@theme(css)
+export class TimePicker extends TimePickerBase<TimePickerProperties> {
+	private _activeIndex: number | undefined;
+	private _endTime: number;
+	private _focused: boolean;
+	private _ignoreBlur: boolean;
+	private _ignoreFocus: boolean;
+	private _inputElement: HTMLInputElement;
+	private _open: boolean;
+	private _options: ReadonlyArray<TimeUnits>;
+	private _startTime: number;
+	private _step: number;
+
+	render(): DNode {
+		const { disabled, formId, invalid, label, readOnly, required } = this.properties;
+		const options = this.renderOptions();
+		const menuId = options ? (<ResultMenuProperties> options.properties).id : '';
+
+		let controls: DNode = v('span', {
+			classes: this.classes(css.controls)
+		}, [
+			this.renderCustomInput(),
+			this.renderButton(menuId as string)
+		]);
+
+		if (label) {
+			controls = w(Label, {
+				bind: this,
+				classes: this.classes(
+					disabled ? css.disabled : null,
+					invalid ? css.invalid : null,
+					readOnly ? css.readOnly : null,
+					required ? css.required : null
+				),
+				formId,
+				label
+			}, [ controls ]);
+		}
+
+		return v('span', {
+			'aria-expanded': this._open ? 'true' : 'false',
+			'aria-haspopup': 'true',
+			'aria-readonly': readOnly ? 'true' : 'false',
+			'aria-required': required ? 'true' : 'false',
```

Comment:
These two should probably be `readOnly/required ? 'true' : null`. I currently don't know of any benefit of including `aria-readonly` and `aria-required` if false.